File: python/src/qcicada/device.py
"""High-level QCicada QRNG device interface."""

from __future__ import annotations

import logging
from dataclasses import replace

from .protocol import (
    CMD_GET_CONFIG, CMD_GET_INFO, CMD_GET_STATISTICS, CMD_GET_STATUS,
    CMD_GET_DEV_PUB_KEY, CMD_GET_DEV_CERTIFICATE, CMD_REBOOT,
    CMD_RESET, CMD_SET_CONFIG, CMD_SIGNED_READ, CMD_START, CMD_STOP,
    RESP_ACK, RESP_NACK, SUCCESS_RESPONSE, PAYLOAD_SIZE,
    MAX_BLOCK_SIZE, SIGNATURE_LEN, PUB_KEY_LEN, CERTIFICATE_LEN,
    build_cmd, build_start_one_shot, build_start_continuous,
    build_signed_read, build_reboot, serialize_config,
    parse_config, parse_info, parse_statistics, parse_status,
)
from .crypto import (
    verify_certificate as _verify_certificate,
    verify_signature as _verify_signature,
    parse_hw_version, parse_serial_int,
)
from .serial import SerialTransport, find_devices
from .types import (
    DeviceConfig, DeviceInfo, DeviceStatistics, DeviceStatus, PostProcess,
    SignedRead,
)

logger = logging.getLogger(__name__)


class QCicadaError(Exception):
    """Raised on unexpected protocol responses or communication failures."""


class QCicada:
    """High-level interface to a QCicada QRNG device.

    Usage::

        with QCicada() as qrng:
            print(qrng.random(32).hex())
    """

    def __init__(self, port: str | None = None, timeout: float = 2.0):
        """Connect to a QCicada device.

        Args:
            port: Serial port path, e.g. ``"/dev/cu.usbserial-DK0HFP4T"``.
                  If None, auto-discovers the first available device.
            timeout: Default read timeout in seconds.

        Raises:
            QCicadaError: If no device is found or the serial port cannot be opened.
        """
        if port is None:
            devices = find_devices()
            if not devices:
                raise QCicadaError('No QCicada device found')
            port = devices[0]
        try:
            self._transport = SerialTransport(port, timeout=timeout)
        except Exception as exc:
            raise QCicadaError(f'Failed to open {port}: {exc}') from exc

    # --- Public API ---

    def get_info(self) -> DeviceInfo:
        """Read device identification (version, serial, hardware)."""
        data = self._command(CMD_GET_INFO)
        if data is None:
            raise QCicadaError('Failed to get device info')
        return parse_info(data)

    def get_status(self) -> DeviceStatus:
        """Read current device status."""
        data = self._command(CMD_GET_STATUS)
        if data is None:
            raise QCicadaError('Failed to get device status')
        return parse_status(data)

    def get_config(self) -> DeviceConfig:
        """Read current device configuration."""
        data = self._command(CMD_GET_CONFIG)
        if data is None:
            raise QCicadaError('Failed to get device config')
        return parse_config(data)

    def set_config(self, config: DeviceConfig) -> None:
        """Write a full device configuration.

        Raises:
            QCicadaError: If the device rejects the configuration.
        """
        payload = serialize_config(config)
        result = self._command(CMD_SET_CONFIG, payload)
        if result is None:
            raise QCicadaError('Failed to set device config (NACK)')

    def get_statistics(self) -> DeviceStatistics:
        """Read generation statistics since last reset."""
        data = self._command(CMD_GET_STATISTICS)
        if data is None:
            raise QCicadaError('Failed to get device statistics')
        return parse_statistics(data)

    def random(self, n: int) -> bytes:
        """Get ``n`` random bytes using one-shot mode.

        Args:
            n: Number of bytes to read (1-65535).

        Raises:
            QCicadaError: If the device fails to respond.
            ValueError: If n is out of range.
        """
        if n == 0:
            return b''
        if not 1 <= n <= 65535:
            raise ValueError(f'n must be 1-65535, got {n}')
        frame = build_start_one_shot(n)
        result = self._command(CMD_START, frame[1:])
        if result is None:
            raise QCicadaError('Failed to start one-shot read')
        # Read the random data
        self._transport.set_timeout(0.5 + 0.0001 * n)
        try:
            data = self._transport.read(n)
        except Exception as exc:
            raise QCicadaError(f'Failed reading random data: {exc}') from exc
        if len(data) != n:
            raise QCicadaError(f'Expected {n} bytes, got {len(data)}')
        return data

    def signed_read(self, n: int) -> SignedRead:
        """Get ``n`` random bytes with a 64-byte cryptographic signature.

        Requires QCicada firmware 5.13+. The signature is produced by the
        device's internal asymmetric key.

        Args:
            n: Number of random bytes to read (1-65535).

        Raises:
            QCicadaError: If the device fails to respond.
            ValueError: If n is out of range.
        """
        if not 1 <= n <= 65535:
            raise ValueError(f'n must be 1-65535, got {n}')
        frame = build_signed_read(n)
        result = self._command(CMD_SIGNED_READ, frame[1:])
        if result is None:
            raise QCicadaError('Failed to start signed read')
        # Read random data + 64-byte signature
        total = n + SIGNATURE_LEN
        self._transport.set_timeout(0.5 + 0.0001 * n)
        try:
            buf = self._transport.read(total)
        except Exception as exc:
            raise QCicadaError(f'Failed reading signed data: {exc}') from exc
        if len(buf) != total:
            raise QCicadaError(f'Expected {total} bytes (data+sig), got {len(buf)}')
        return SignedRead(data=buf[:n], signature=buf[n:])

    def start_continuous(self) -> None:
        """Start continuous random data generation.

        After calling this, use :meth:`read_continuous` to read streaming data.
        Call :meth:`stop` to end continuous mode.

        Raises:
            QCicadaError: If the device rejects the command.
        """
        frame = build_start_continuous()
        result = self._command(CMD_START, frame[1:])
        if result is None:
            raise QCicadaError('Failed to start continuous mode')

    def read_continuous(self, n: int) -> bytes:
        """Read bytes from an active continuous mode stream.

        Call :meth:`start_continuous` first. Returns exactly ``n`` bytes.

        Args:
            n: Number of bytes to read.

        Raises:
            QCicadaError: If the read fails or times out.
        """
        if n == 0:
            return b''
        self._transport.set_timeout(0.5 + 0.0001 * n)
        try:
            data = self._transport.read(n)
        except Exception as exc:
            raise QCicadaError(f'Failed reading continuous data: {exc}') from exc
        if len(data) != n:
            raise QCicadaError(f'Expected {n} continuous bytes, got {len(data)}')
        return data

    def get_dev_pub_key(self) -> bytes:
        """Retrieve the device's ECDSA P-256 public key (64 bytes: x || y).

        Requires QCicada firmware with certificate support.
        """
        data = self._command(CMD_GET_DEV_PUB_KEY)
        if data is None:
            raise QCicadaError('Failed to get device public key (NACK)')
        if len(data) != PUB_KEY_LEN:
            raise QCicadaError(
                f'Expected {PUB_KEY_LEN} byte public key, got {len(data)}'
            )
        return data

    def get_dev_certificate(self) -> bytes:
        """Retrieve the device certificate (64 bytes: ECDSA r || s).

        This is the CA's signature over the device's identity (hw version,
        serial number, and public key).
        """
        data = self._command(CMD_GET_DEV_CERTIFICATE)
        if data is None:
            raise QCicadaError('Failed to get device certificate (NACK)')
        if len(data) != CERTIFICATE_LEN:
            raise QCicadaError(
                f'Expected {CERTIFICATE_LEN} byte certificate, got {len(data)}'
            )
        return data

    def get_verified_pub_key(self, ca_pub_key: bytes) -> bytes:
        """Retrieve and verify the device's public key against a CA public key.

        Fetches the device's info, public key, and certificate, then verifies
        the certificate chain. Returns the verified public key on success.

        Args:
            ca_pub_key: 64 bytes (x || y) of the Certificate Authority's public key.

        Raises:
            QCicadaError: If verification fails or device communication fails.
        """
        info = self.get_info()
        dev_pub_key = self.get_dev_pub_key()
        certificate = self.get_dev_certificate()

        hw_ver = parse_hw_version(info.hw_info)
        if hw_ver is None:
            raise QCicadaError(
                f"Cannot parse hardware version from '{info.hw_info}'"
            )
        hw_major, hw_minor = hw_ver

        serial_int = parse_serial_int(info.serial)
        if serial_int is None:
            raise QCicadaError(
                f"Cannot parse serial number from '{info.serial}'"
            )

        valid = _verify_certificate(
            ca_pub_key, dev_pub_key, certificate,
            hw_major, hw_minor, serial_int,
        )
        if not valid:
            raise QCicadaError('Device certificate verification failed')
        return dev_pub_key

    def signed_read_verified(self, n: int, device_pub_key: bytes) -> SignedRead:
        """Perform a signed read and verify the signature.

        Args:
            n: Number of random bytes to read (1-65535).
            device_pub_key: 64 bytes (x || y) of the device's verified public key.

        Returns:
            Verified :class:`SignedRead` with data and signature.

        Raises:
            QCicadaError: If verification fails.
        """
        result = self.signed_read(n)
        valid = _verify_signature(device_pub_key, result.data, result.signature)
        if not valid:
            raise QCicadaError('Signed read signature verification failed')
        return result

    def reboot(self) -> None:
        """Reboot the device.

        The device will disconnect and reconnect — you must re-open the
        connection after calling this.
        """
        frame = build_reboot()
        self._transport.flush()
        try:
            self._transport.write(frame)
        except IOError:
            pass  # Device may disconnect immediately
        # Try to read optional response
        try:
            self._transport.set_timeout(0.5)
            self._transport.read(1)
        except Exception:
            pass

    def fill_bytes(self, buf: bytearray) -> None:
        """Fill a buffer with random bytes, chunking as needed.

        Handles reads larger than the 65535-byte protocol limit automatically.

        Args:
            buf: Mutable buffer to fill with random data.
        """
        offset = 0
        while offset < len(buf):
            chunk = min(len(buf) - offset, 65535)
            data = self.random(chunk)
            buf[offset:offset + len(data)] = data
            offset += len(data)

    def set_postprocess(self, mode: PostProcess) -> None:
        """Change post-processing mode, preserving other config settings.

        Raises:
            QCicadaError: If the device rejects the change.
        """
        config = self.get_config()
        config = replace(config, postprocess=mode)
        self.set_config(config)

    def reset(self) -> None:
        """Reset the device (restarts startup test and clears statistics).

        Raises:
            QCicadaError: If the device rejects the reset.
        """
        result = self._command(CMD_RESET)
        if result is None:
            raise QCicadaError('Failed to reset device (NACK)')

    def stop(self) -> None:
        """Send STOP command to halt any active generation."""
        self._command(CMD_STOP)

    def close(self) -> None:
        """Close the serial connection."""
        self._transport.close()

    def __enter__(self) -> QCicada:
        return self

    def __exit__(self, *args) -> None:
        self.close()

    # --- Internal protocol handling ---

    def _command(self, cmd_code: bytes, payload: bytes | None = None):
        """Send a command and read the response.

        Returns the response payload bytes, True (for empty-payload success),
        or None on NACK.

        Raises QCicadaError on write failures or unexpected response bytes.
        """
        if cmd_code not in SUCCESS_RESPONSE:
            raise ValueError(f'Unknown command code: {cmd_code.hex()}')

        self._transport.flush()
        frame = build_cmd(cmd_code, payload)

        try:
            self._transport.write(frame)
        except IOError as exc:
            raise QCicadaError(f'Write failed: {exc}') from exc

        # STOP command: drain buffer, find ACK near end
        if cmd_code == CMD_STOP:
            return self._handle_stop()

        # Normal command: read 1-byte response code
        try:
            self._transport.set_timeout(3)
            resp = self._transport.read(1)
            if len(resp) == 0:
                return None
        except Exception as exc:
            raise QCicadaError(f'Read failed: {exc}') from exc

        expected = SUCCESS_RESPONSE[cmd_code]
        if resp == expected:
            expected_size = PAYLOAD_SIZE[expected]
            if expected_size == 0:
                return True
            self._transport.set_timeout(max(0.5, 0.001 * expected_size))
            try:
                resp_payload = self._transport.read(expected_size)
            except Exception as exc:
                raise QCicadaError(f'Read payload failed: {exc}') from exc
            if len(resp_payload) != expected_size:
                return None
            return resp_payload
        elif resp == RESP_NACK:
            return None
        else:
            raise QCicadaError(f'Unexpected response byte: 0x{resp.hex()}')

    def _handle_stop(self):
        """Handle STOP command response: drain pipe and find ACK."""
        ack_payload_size = PAYLOAD_SIZE[RESP_ACK]
        drain_size = MAX_BLOCK_SIZE * 2 + ack_payload_size + 1

        self._transport.set_timeout(0.5)
        for _ in range(2):
            resp = self._transport.read(drain_size)
            if len(resp) == 1 and resp == RESP_NACK:
                return None
            if len(resp) < ack_payload_size + 1:
                return None
            # Check for ACK at expected position from end
            ack_pos = len(resp) - 1 - ack_payload_size
            if resp[ack_pos] == RESP_ACK[0]:
                return True
        return None

File: src/device.rs
//! High-level QCicada QRNG device interface.

use std::io;
use std::time::Duration;

use crate::protocol::*;
use crate::serial::{find_devices, SerialTransport};
use crate::types::*;
use crate::QCicadaError;

/// High-level interface to a QCicada QRNG device.
///
/// ```no_run
/// use qcicada::QCicada;
///
/// let mut qrng = QCicada::open(None, None)?;
/// let bytes = qrng.random(32)?;
/// println!("{:02x?}", &bytes);
/// qrng.close();
/// # Ok::<(), qcicada::QCicadaError>(())
/// ```
pub struct QCicada {
    transport: SerialTransport,
}

impl QCicada {
    /// Connect to a QCicada device.
    ///
    /// - `port`: Serial port path. If `None`, auto-discovers the first available device.
    /// - `timeout`: Default read timeout. If `None`, uses 2 seconds.
    pub fn open(port: Option<&str>, timeout: Option<Duration>) -> Result<Self, QCicadaError> {
        let timeout = timeout.unwrap_or(Duration::from_secs(2));

        let port_name = match port {
            Some(p) => p.to_string(),
            None => {
                let devices = find_devices();
                devices
                    .into_iter()
                    .next()
                    .ok_or(QCicadaError::NoDevice)?
            }
        };

        let transport = SerialTransport::open(&port_name, timeout)?;
        Ok(Self { transport })
    }

    /// Read device identification (version, serial, hardware).
    pub fn get_info(&mut self) -> Result<DeviceInfo, QCicadaError> {
        let data = self.command(CMD_GET_INFO, None)?
            .ok_or(QCicadaError::Protocol("No response to GET_INFO".into()))?;
        parse_info(&data)
    }

    /// Read current device status.
    pub fn get_status(&mut self) -> Result<DeviceStatus, QCicadaError> {
        let data = self.command(CMD_GET_STATUS, None)?
            .ok_or(QCicadaError::Protocol("No response to GET_STATUS".into()))?;
        parse_status(&data)
    }

    /// Read current device configuration.
    pub fn get_config(&mut self) -> Result<DeviceConfig, QCicadaError> {
        let data = self.command(CMD_GET_CONFIG, None)?
            .ok_or(QCicadaError::Protocol("No response to GET_CONFIG".into()))?;
        parse_config(&data)
    }

    /// Write a full device configuration.
    pub fn set_config(&mut self, config: &DeviceConfig) -> Result<(), QCicadaError> {
        let payload = serialize_config(config);
        self.command(CMD_SET_CONFIG, Some(&payload))?
            .ok_or(QCicadaError::Protocol("NACK on SET_CONFIG".into()))?;
        Ok(())
    }

    /// Read generation statistics since last reset.
    pub fn get_statistics(&mut self) -> Result<DeviceStatistics, QCicadaError> {
        let data = self.command(CMD_GET_STATISTICS, None)?
            .ok_or(QCicadaError::Protocol("No response to GET_STATISTICS".into()))?;
        parse_statistics(&data)
    }

    /// Get `n` random bytes using one-shot mode.
    ///
    /// Maximum 65535 bytes per call (protocol limit).
    pub fn random(&mut self, n: u16) -> Result<Vec<u8>, QCicadaError> {
        if n == 0 {
            return Ok(Vec::new());
        }
        let frame = build_start_one_shot(n);
        // Send: CMD_START is frame[0], payload is frame[1..]
        self.command(CMD_START, Some(&frame[1..]))?
            .ok_or(QCicadaError::Protocol("NACK on START one-shot".into()))?;

        // Read the random data
        let timeout_ms = 500 + (n as u64) / 10;
        self.transport
            .set_timeout(Duration::from_millis(timeout_ms))?;
        let data = self.transport.read(n as usize)?;
        if data.len() != n as usize {
            return Err(QCicadaError::Protocol(format!(
                "Expected {} random bytes, got {}",
                n,
                data.len()
            )));
        }
        Ok(data)
    }

    /// Get `n` random bytes with a 64-byte cryptographic signature.
    ///
    /// Requires QCicada firmware 5.13+. The signature is produced by the device's
    /// internal asymmetric key. See device documentation for the public key.
    pub fn signed_read(&mut self, n: u16) -> Result<SignedRead, QCicadaError> {
        if n == 0 {
            return Err(QCicadaError::Protocol(
                "signed_read requires at least 1 byte".into(),
            ));
        }
        let frame = build_signed_read(n);
        self.command(CMD_SIGNED_READ, Some(&frame[1..]))?
            .ok_or(QCicadaError::Protocol("NACK on SIGNED_READ".into()))?;

        // Read random data + 64-byte signature
        let total = n as usize + SIGNATURE_LEN;
        let timeout_ms = 500 + (n as u64) / 10;
        self.transport
            .set_timeout(Duration::from_millis(timeout_ms))?;
        let buf = self.transport.read(total)?;
        if buf.len() != total {
            return Err(QCicadaError::Protocol(format!(
                "Expected {} bytes (data+sig), got {}",
                total,
                buf.len()
            )));
        }
        Ok(SignedRead {
            data: buf[..n as usize].to_vec(),
            signature: buf[n as usize..].to_vec(),
        })
    }

    /// Start continuous random data generation.
    ///
    /// After calling this, use [`read_continuous`] to read streaming data.
    /// Call [`stop`] to end continuous mode.
    pub fn start_continuous(&mut self) -> Result<(), QCicadaError> {
        let frame = build_start_continuous();
        self.command(CMD_START, Some(&frame[1..]))?
            .ok_or(QCicadaError::Protocol("NACK on START continuous".into()))?;
        Ok(())
    }

    /// Read bytes from an active continuous mode stream.
    ///
    /// Call [`start_continuous`] first. Returns exactly `n` bytes or an error.
    pub fn read_continuous(&mut self, n: usize) -> Result<Vec<u8>, QCicadaError> {
        if n == 0 {
            return Ok(Vec::new());
        }
        let timeout_ms = 500 + (n as u64) / 10;
        self.transport
            .set_timeout(Duration::from_millis(timeout_ms))?;
        let data = self.transport.read(n)?;
        if data.len() != n {
            return Err(QCicadaError::Protocol(format!(
                "Expected {} continuous bytes, got {}",
                n,
                data.len()
            )));
        }
        Ok(data)
    }

    /// Retrieve the device's ECDSA P-256 public key (64 bytes: x || y).
    ///
    /// Requires QCicada firmware with certificate support.
    pub fn get_dev_pub_key(&mut self) -> Result<Vec<u8>, QCicadaError> {
        let data = self
            .command(CMD_GET_DEV_PUB_KEY, None)?
            .ok_or(QCicadaError::Protocol("NACK on GET_DEV_PUB_KEY".into()))?;
        if data.len() != PUB_KEY_LEN {
            return Err(QCicadaError::Protocol(format!(
                "Expected {} byte public key, got {}",
                PUB_KEY_LEN,
                data.len()
            )));
        }
        Ok(data)
    }

    /// Retrieve the device certificate (64 bytes: ECDSA r || s).
    ///
    /// This is the CA's signature over the device's identity (hw version,
    /// serial number, and public key).
    pub fn get_dev_certificate(&mut self) -> Result<Vec<u8>, QCicadaError> {
        let data = self
            .command(CMD_GET_DEV_CERTIFICATE, None)?
            .ok_or(QCicadaError::Protocol("NACK on GET_DEV_CERTIFICATE".into()))?;
        if data.len() != CERTIFICATE_LEN {
            return Err(QCicadaError::Protocol(format!(
                "Expected {} byte certificate, got {}",
                CERTIFICATE_LEN,
                data.len()
            )));
        }
        Ok(data)
    }

    /// Retrieve and verify the device's public key against a CA public key.
    ///
    /// Fetches the device's info, public key, and certificate, then verifies
    /// the certificate chain. Returns the verified public key on success.
    ///
    /// # Arguments
    /// - `ca_pub_key`: 64 bytes (x || y) of the Certificate Authority's public key.
    pub fn get_verified_pub_key(
        &mut self,
        ca_pub_key: &[u8],
    ) -> Result<Vec<u8>, QCicadaError> {
        let info = self.get_info()?;
        let dev_pub_key = self.get_dev_pub_key()?;
        let certificate = self.get_dev_certificate()?;

        let (hw_major, hw_minor) = crate::protocol::parse_hw_version(&info.hw_info)
            .ok_or_else(|| {
                QCicadaError::Protocol(format!(
                    "Cannot parse hardware version from '{}'",
                    info.hw_info
                ))
            })?;
        let serial_int = crate::protocol::parse_serial_int(&info.serial).ok_or_else(|| {
            QCicadaError::Protocol(format!(
                "Cannot parse serial number from '{}'",
                info.serial
            ))
        })?;

        let valid = crate::crypto::verify_certificate(
            ca_pub_key,
            &dev_pub_key,
            &certificate,
            hw_major,
            hw_minor,
            serial_int,
        )
        .map_err(|e| QCicadaError::Protocol(format!("Certificate verification error: {e}")))?;

        if !valid {
            return Err(QCicadaError::Protocol(
                "Device certificate verification failed".into(),
            ));
        }
        Ok(dev_pub_key)
    }

    /// Perform a signed read and verify the signature against a known device public key.
    ///
    /// Returns the verified [`SignedRead`] on success. Fails if verification fails.
    ///
    /// # Arguments
    /// - `n`: Number of random bytes to read.
    /// - `device_pub_key`: 64 bytes (x || y) of the device's verified public key.
    pub fn signed_read_verified(
        &mut self,
        n: u16,
        device_pub_key: &[u8],
    ) -> Result<SignedRead, QCicadaError> {
        let result = self.signed_read(n)?;

        let valid = crate::crypto::verify_signature(device_pub_key, &result.data, &result.signature)
            .map_err(|e| {
                QCicadaError::Protocol(format!("Signature verification error: {e}"))
            })?;

        if !valid {
            return Err(QCicadaError::Protocol(
                "Signed read signature verification failed".into(),
            ));
        }
        Ok(result)
    }

    /// Reboot the device.
    ///
    /// Sends the QCicada-specific reboot command. The device will disconnect
    /// and reconnect — you must re-open the connection after calling this.
    pub fn reboot(&mut self) -> Result<(), QCicadaError> {
        let frame = build_reboot();
        // Send the full frame (command + magic) directly
        self.transport.flush()?;
        self.transport.write(&frame)?;
        // Read optional response — device may disconnect immediately
        self.transport.set_timeout(Duration::from_millis(500))?;
        let _ = self.transport.read(1);
        Ok(())
    }

    /// Change post-processing mode, preserving other config settings.
    pub fn set_postprocess(&mut self, mode: PostProcess) -> Result<(), QCicadaError> {
        let mut config = self.get_config()?;
        config.postprocess = mode;
        self.set_config(&config)
    }

    /// Reset the device (restarts startup test and clears statistics).
    pub fn reset(&mut self) -> Result<(), QCicadaError> {
        self.command(CMD_RESET, None)?
            .ok_or(QCicadaError::Protocol("NACK on RESET".into()))?;
        Ok(())
    }

    /// Send STOP command to halt any active generation.
    pub fn stop(&mut self) -> Result<(), QCicadaError> {
        self.command(CMD_STOP, None)?;
        Ok(())
    }

    /// Fill a buffer with random bytes, chunking as needed for the protocol limit.
    pub fn fill_bytes(&mut self, buf: &mut [u8]) -> Result<(), QCicadaError> {
        let mut offset = 0;
        while offset < buf.len() {
            let chunk = (buf.len() - offset).min(u16::MAX as usize) as u16;
            let data = self.random(chunk)?;
            buf[offset..offset + data.len()].copy_from_slice(&data);
            offset += data.len();
        }
        Ok(())
    }

    /// Close the serial connection.
    pub fn close(self) {
        drop(self);
    }

    // --- Internal protocol handling ---

    /// Send a command and read the response.
    ///
    /// Returns `Ok(Some(payload))` on success with payload,
    /// `Ok(Some(empty vec))` on success without payload (ACK with no data expected beyond status),
    /// or `Ok(None)` on NACK.
    fn command(
        &mut self,
        cmd_code: u8,
        payload: Option<&[u8]>,
    ) -> Result<Option<Vec<u8>>, QCicadaError> {
        let expected = expected_response(cmd_code)
            .ok_or_else(|| QCicadaError::Protocol(format!("Unknown command: {cmd_code:#04x}")))?;

        self.transport.flush()?;

        let frame = build_cmd(cmd_code, payload);
        self.transport.write(&frame)?;

        // STOP command: drain buffer, find ACK near end
        if cmd_code == CMD_STOP {
            return self.handle_stop();
        }

        // Read 1-byte response code
        self.transport.set_timeout(Duration::from_secs(3))?;
        let resp = self.transport.read(1)?;
        if resp.is_empty() {
            return Ok(None);
        }

        if resp[0] == expected {
            let size = payload_size(expected);
            if size == 0 {
                return Ok(Some(Vec::new()));
            }
            let timeout_ms = (size as u64).max(500);
            self.transport
                .set_timeout(Duration::from_millis(timeout_ms))?;
            let resp_payload = self.transport.read(size)?;
            if resp_payload.len() != size {
                return Ok(None);
            }
            Ok(Some(resp_payload))
        } else if resp[0] == RESP_NACK {
            Ok(None)
        } else {
            Err(QCicadaError::Protocol(format!(
                "Unexpected response byte: {:#04x}",
                resp[0]
            )))
        }
    }

    /// Handle STOP command response: drain pipe and find ACK.
    fn handle_stop(&mut self) -> Result<Option<Vec<u8>>, QCicadaError> {
        let drain_size = MAX_BLOCK_SIZE * 2 + PAYLOAD_ACK + 1;

        self.transport.set_timeout(Duration::from_millis(500))?;

        for _ in 0..2 {
            let resp = self.transport.read(drain_size)?;
            if resp.len() == 1 && resp[0] == RESP_NACK {
                return Ok(None);
            }
            if resp.len() < PAYLOAD_ACK + 1 {
                return Ok(None);
            }
            let ack_pos = resp.len() - 1 - PAYLOAD_ACK;
            if resp[ack_pos] == RESP_ACK {
                return Ok(Some(Vec::new()));
            }
        }
        Ok(None)
    }
}

impl io::Read for QCicada {
    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
        if buf.is_empty() {
            return Ok(0);
        }
        let n = buf.len().min(u16::MAX as usize) as u16;
        let data = self
            .random(n)
            .map_err(io::Error::other)?;
        buf[..data.len()].copy_from_slice(&data);
        Ok(data.len())
    }
}
